ckanext/canada/logic.py

```diff
@@ -185,3 +194,88 @@ def datastore_create_temp_user_table(context):
+
+    Canada Fork: extends the resource_view_show action method
+    to prevent showing datatable_views for invalid and inactive Resources
+
+    We need this method to 404 the resource_view page (e.g. viewing in full screen)
+
+    If a user is logged in, we want them to be able to see and edit the datatables_view
+    views still. We will just add a key to the view dict to be used within templates for visuals.
+    """
+    view_dict = up_func(context, data_dict)
+    if view_dict.get('view_type') == 'datatables_view':
+        # at this point, the core function has been called, calling resource_show
+        # and adding the Resource Object into the context
+        res_dict = context.get('resource').__dict__
+        res_extras = res_dict.get('extras', {})
+        site_user = get_action('get_site_user')({'ignore_auth': True}, {})['name']
+
+        if not res_extras.get('datastore_active', False) or \
+        res_extras.get('validation_status') != 'success':
+            # if the Resource is not active in the DataStore or has not
+            # passed validation, we do not want to show its datatables_view views
+
+            if not g.user:
+                # only raise if a user is not logged in
+                raise ObjectNotFound
+
+            if g.user != site_user:
+                # only add key/value if not system process
+                view_dict['canada_disabled_view'] = True
+
+    return view_dict
+
+
+@chained_action
+def canada_resource_view_list(up_func, context, data_dict):
+    """
+    Return the list of resource views for a particular resource.
+
+    :param id: the id of the resource
+    :type id: string
+
+    :rtype: list of dictionaries.
+
+    Canada Fork: extends the resource_view_list action method
+    to prevent showing datatable_views for invalid and inactive Resources
+
+    This will delete any datatables_view from the list of views.
+
+    If a user is logged in, we want them to be able to see and edit the datatables_view
+    views still. We will just add a key to the view dict to be used within templates for visuals.
+    """
+    view_list = up_func(context, data_dict)
+    # at this point, the core function has been called, calling resource_show
+    # and adding the Resource Object into the context
+    res_dict = context.get('resource').__dict__
+    res_extras = res_dict.get('extras', {})
+    site_user = get_action('get_site_user')({'ignore_auth': True}, {})['name']
+    for i, view_dict in enumerate(view_list):
+
+        if view_dict.get('view_type') == 'datatables_view' and \
+        ( not res_extras.get('datastore_active', False) or
+          res_extras.get('validation_status') != 'success'):
+            # if the Resource is not active in the DataStore or has not
+            # passed validation, we do not want to show its datatables_view views
+
+            if not g.user:
+                # only remove from the list if a user is not logged in
+                del view_list[i]
```

Comment:
it's unsafe to modify a list while iterating over it

Reply:
save index numbers to a new list, and then bisect after the loop is done??

Reply:
okay, just saved index numbers and then removed after loop